### PR TITLE
#2039 add cash register balance

### LINKED ISSUE #2039: Add cash register balance
## Is your feature request related to a problem? Please describe.

Add a widget to the `CashRegisterPage` which renders the current cash balance. 

## Describe the solution you'd like

See above.

## Implementation

Should be relatively straightforward, e.g. a selector to derive the balance from page transactions, a simple `Text` component to display the value (most real estate is probably the the top-left of the UI).

## Describe alternatives you've considered

N/A.

## Additional context

N/A.

### PR DESCRIPTION
**Pull request by wlthomson**
Fixes #2039.

## Change summary

Adds current balance to cash register page.

## Testing

Steps to reproduce or otherwise test the changes of this PR:

- [ ] Cash register display correct balance (`sum(receipts)-sum(payments)`).

### Related areas to think about

Top-left UI isn't particularly nice atm, but thinking that after addition of a searchbar etc. will start looking better.

### CODE REVIEW

File: src/selectors/cashRegister.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * Derives current balance from cash register transactions.
+ * @return {Number}
+ */
+export const selectTransactions = createSelector(
+  [pageStateSelector],
+  pageState => pageState.backingData
+);
+
+export const selectReceipts = createSelector([selectTransactions], transactions =>
+  transactions.filter(({ type }) => type === 'receipt')
```
Comment (joshxg) [inline src/selectors/cashRegister.js]: you can use `filtered` if your backing data is just all `Transaction`  - pretty sure it'd be faster regardless

File: src/selectors/cashRegister.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+);
+
+export const selectPayments = createSelector([selectTransactions], transactions =>
+  transactions.filter(({ type }) => type === 'payment')
+);
+
+export const selectReceiptsTotal = createSelector([selectReceipts], receipts =>
+  receipts.reduce((acc, { total }) => acc + total, 0)
+);
```
Comment (joshxg) [inline src/selectors/cashRegister.js]: `receipts.reduce(acc, {total}) => currency.add(total), currency(0))`;

File: src/selectors/cashRegister.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+export const selectPaymentsTotal = createSelector([selectPayments], payments =>
+  payments.reduce((acc, { total }) => acc + total, 0)
+);
```
Comment (joshxg) [inline src/selectors/cashRegister.js]: `payments.reduce(acc, {total}) => currency.add(total), currency(0))`;

File: src/selectors/cashRegister.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+export const selectBalance = createSelector(
+  [selectReceiptsTotal, selectPaymentsTotal],
+  (receiptsTotal, paymentsTotal) => currency(receiptsTotal - paymentsTotal).format(false)
+);
```
Comment (joshxg) [inline src/selectors/cashRegister.js]: `... => receiptsTotal.subtract(paymentsTotal).format()`;

File: src/selectors/cashRegister.js (hunk 1 above)
Comment (wlthomson) [reply src/selectors/cashRegister.js]: Unfortunately not possible, this is where the realm issue occurs 😢 .

File: src/selectors/cashRegister.js (hunk 3 above)
Comment (wlthomson) [reply src/selectors/cashRegister.js]: 5e850fc85216fed6848878eaf66841c8dc25fc1b.

File: src/selectors/cashRegister.js (hunk 2 above)
Comment (wlthomson) [reply src/selectors/cashRegister.js]: 5e850fc85216fed6848878eaf66841c8dc25fc1b.

File: src/selectors/cashRegister.js (hunk 4 above)
Comment (wlthomson) [reply src/selectors/cashRegister.js]: 5e850fc85216fed6848878eaf66841c8dc25fc1b.

File: src/pages/CashRegisterPage.js [hunk 5]
```diff
@@ -9,21 +9,37 @@ import PropTypes from 'prop-types';
...
+  keyExtractor,
+  modalKey,
+  columns,
+  currentBalance,
+  onNewCashTransaction,
+  onCloseModal,
+  onAddCashTransaction,
+}) => {
+  // eslint-disable-next-line no-unused-vars
+  const getCallback = _ => null;
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply src/pages/CashRegisterPage.js]: 1548e1b688dcbe774209ee91916f0b735061e48b.

File: src/pages/CashRegisterPage.js [hunk 6]
```diff
@@ -9,21 +9,37 @@ import PropTypes from 'prop-types';
...
+
+  // eslint-disable-next-line no-unused-vars
+  const onPressRow = _ => null;
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply src/pages/CashRegisterPage.js]: 0efd157239c6e58b3c7c7fb0f154326974f318f3.

File: src/pages/CashRegisterPage.js [hunk 7]
```diff
@@ -70,8 +86,14 @@ export const CashRegister = ({ dispatch, data, dataState, sortKey, keyExtractor,
   return (
     <DataTablePageView>
       <View style={pageTopSectionContainer}>
-        <View style={pageTopLeftSectionContainer} />
-        <View style={pageTopRightSectionContainer}>
+        <View style={{ ...pageTopLeftSectionContainer, ...{ flex: 1 } }}>
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply src/pages/CashRegisterPage.js]: Interestingly `FlexView` component had inconsistent behaviour to inline (can demo if you're interested). Tidied up with `useMemo`. f7a6dd027a7be3c80c0ec87cf477bb1c6d185513.

File: src/pages/CashRegisterPage.js [hunk 8]
```diff
@@ -102,20 +124,26 @@ const mapDispatchToProps = (dispatch, ownProps) => ({
 const mapStateToProps = state => {
   const { pages } = state;
   const { cashRegister } = pages;
-  return cashRegister;
+
+  const currentBalance = selectBalance(cashRegister);
```
[the remark this answers is not in the document]
Comment (wlthomson) [reply src/pages/CashRegisterPage.js]: Definitely, pretty much the whole point. Not sure what I was thinking here. aed61867f8ecf5210976705ec32b6d69ac9787e2.

File: src/selectors/cashRegister.js (hunk 1 above)
Comment (wlthomson) [reply src/selectors/cashRegister.js]: Update: whether realm bug or something else theoretical now as `transactions` is not longer guaranteed to be a realm collection (workaround for rerender during transition).

**Pull request merged**